### PR TITLE
[ApiBundle][Statystics] Add validation

### PR DESCRIPTION
**Pull request by Wojdylak**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.13
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | #15765
| License         | MIT

### CODE REVIEW

File: tests/Api/Admin/StatisticsTest.php [hunk 1]
```diff
@@ -169,7 +225,7 @@ public function it_returns_a_bad_request_status_code_if_any_of_required_paramete
             server: $this->headerBuilder()->withAdminUserAuthorization('api@example.com')->build(),
         );
 
-        $this->assertResponseCode($this->client->getResponse(), Response::HTTP_BAD_REQUEST);
+        $this->assertResponseCode($this->client->getResponse(), Response::HTTP_UNPROCESSABLE_ENTITY);
```
Comment (NoResponseMate) [inline tests/Api/Admin/StatisticsTest.php]: Would be nice to adjust the name of this test.

File: tests/Api/Admin/StatisticsTest.php [hunk 2]
```diff
@@ -150,6 +150,62 @@ public function it_returns_a_not_found_status_code_if_channel_does_not_exist():
...
+        $this->assertResponseViolations(
+            $this->client->getResponse(),
+            [
+                [
+                    'propertyPath' => '',
+                    'message' => 'The start date must be earlier than the end date.',
+                ],
+            ],
+        );
+    }
```
Comment (NoResponseMate) [inline tests/Api/Admin/StatisticsTest.php]: You can merge this with the one below and use a data provider for both cases.
Also, while you're at it, the `::invalidPeriods()` method is unused and can be either adjusted for this test or removed.

File: src/Sylius/Bundle/CoreBundle/DependencyInjection/Compiler/CheckStatisticsOrdersTotalsProviderTypePass.php [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+        $intervalsType = $container->getParameter('sylius_core.orders_statistics.intervals_map');
+        $ordersTotalsProviderTypes = [];
+
+        foreach ($container->findTaggedServiceIds('sylius.statistics.orders_totals_provider') as $id => $attributes) {
+            foreach ($attributes as $attribute) {
+                if (!isset($attribute['type'])) {
+                    throw new \InvalidArgumentException('Tagged orders totals providers need to have `type` attribute.');
+                }
+                $ordersTotalsProviderTypes[] = $attribute['type'];
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/CoreBundle/DependencyInjection/Compiler/CheckStatisticsOrdersTotalsProviderTypePass.php]:
```suggestion
                }

                $ordersTotalsProviderTypes[] = $attribute['type'];
```

File: src/Sylius/Bundle/CoreBundle/DependencyInjection/Compiler/CheckStatisticsOrdersTotalsProviderTypePass.php [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+declare(strict_types=1);
+
+namespace Sylius\Bundle\CoreBundle\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+
+final class CheckStatisticsOrdersTotalsProviderTypePass implements CompilerPassInterface
+{
+
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/CoreBundle/DependencyInjection/Compiler/CheckStatisticsOrdersTotalsProviderTypePass.php]:
```suggestion
```

File: src/Sylius/Bundle/CoreBundle/DependencyInjection/Compiler/CheckStatisticsOrdersTotalsProviderTypePass.php [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+                }
+                $ordersTotalsProviderTypes[] = $attribute['type'];
+            }
+        }
+
+        foreach ($intervalsType as $type => $interval) {
+            if (!in_array($type, $ordersTotalsProviderTypes, true)) {
+                throw new \InvalidArgumentException(sprintf('There is no orders totals provider for interval type "%s"', $type));
+            }
+
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/CoreBundle/DependencyInjection/Compiler/CheckStatisticsOrdersTotalsProviderTypePass.php]:
```suggestion
```

File: src/Sylius/Bundle/CoreBundle/DependencyInjection/Compiler/CheckStatisticsOrdersTotalsProviderTypePass.php [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+
+final class CheckStatisticsOrdersTotalsProviderTypePass implements CompilerPassInterface
+{
+
+    public function process(ContainerBuilder $container): void
+    {
+        if (!$container->hasParameter('sylius_core.orders_statistics.intervals_map')) {
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/CoreBundle/DependencyInjection/Compiler/CheckStatisticsOrdersTotalsProviderTypePass.php:23]: Something that completely slipped my mind, but we should add some validation in the `Configuration` class as well.

`cannotBeEmpty` for both `interval` and `period_format` nodes, and custom validation for the interval which will check that it's a duration string.

File: src/Sylius/Bundle/CoreBundle/DependencyInjection/Compiler/CheckStatisticsOrdersTotalsProviderTypePass.php [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+            return;
+        }
+
+        $intervalsType = $container->getParameter('sylius_core.orders_statistics.intervals_map');
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/DependencyInjection/Compiler/CheckStatisticsOrdersTotalsProviderTypePass.php]:
```suggestion
        $intervalsTypes = $container->getParameter('sylius_core.orders_statistics.intervals_map');
```

**Review by NoResponseMate — DISMISSED** [review]

File: src/Sylius/Bundle/CoreBundle/DependencyInjection/Configuration.php [hunk 8]
```diff
@@ -68,8 +68,24 @@ public function getConfigTreeBuilder(): TreeBuilder
...
+                                    ->scalarNode('interval')
+                                        ->cannotBeEmpty()
+                                        ->validate()
+                                            ->ifTrue(
+                                                function (mixed $interval) {
+                                                    try {
+                                                        new \DateInterval($interval);
+
+                                                        return false;
+                                                    } catch (\Exception $e) {
```
Comment (oallain) [inline src/Sylius/Bundle/CoreBundle/DependencyInjection/Configuration.php:80]: Can be written
```suggestion
                                                    } catch (\Exception) {
```

File: src/Sylius/Bundle/CoreBundle/Tests/DependencyInjection/Compiler/CheckStatisticsOrdersTotalsProviderTypePassTest.php [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+        $this->setParameter('sylius_core.orders_statistics.intervals_map', ['daily' => 'Daily', 'monthly' => 'Monthly']);
+        $this->setDefinition(
+            'sylius.registry.statistics_orders_totals_provider',
+            (new Definition())
+                ->addTag('sylius.statistics.orders_totals_provider', ['type' => 'daily'])
+                ->addTag('sylius.statistics.orders_totals_provider', ['type' => 'monthly']),
+        );
+
+        $this->compile();
+        $this->assertTrue(true, 'Compiler pass completed without throwing an exception.');
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/CoreBundle/Tests/DependencyInjection/Compiler/CheckStatisticsOrdersTotalsProviderTypePassTest.php]:
```suggestion
```
Makes no sense; Guessing it's only here to shut up unit throwing hissy fits about a risky test or some other crap. Just assert after compilation if the service and parameter are there.

**Review by NoResponseMate — APPROVED** [review]

**Comment by NoResponseMate** [comment]
Thank you, @Wojdylak!

**Pull request merged**